enhancements/network/ovn-hypershift-live-migration.md

```diff
@@ -0,0 +1,303 @@
+
+The bridge binding is able to expose the pod ip to the VM as expected by hypershift but kubevirt does not support live migration with it, for kubevirt to support live migration with bridge binding 
+at primary interface the OVN kubernetes CNI need to do the following:
+- not set the IP at pod netns
+- not expect the IP to be on the pod netns
+
+
+### Workflow Description
+
+The gist of the design is:
+
+1. Hypershift kubevirt workers attach to ovn-k node switches but CNI does not set IP addresses in virt-launcher pods netns
+2. DHCP is configured at VMs logical switch ports to deliver the address to VMs
+3. A point to point routing is used so one node's subnet ip can be routed from different node
+3. The VM's gateway ip and mac are independent of the node they are running on using proxy arp
+4. Current OVN arp proxy feature is missing following features: configure MAC and IPv6, they are being implemented [here](https://github.com/ovn-org/ovn/pull/176)
+
+
+### API Extensions
+
+n/a
+
+## Design Details
+
+### OVN-Kubernetes 
+
+#### Topology
+
+**Point to point routing:**
+
+Do point to point routing with a policy for outbound traffic and a static route for
+inbound, with this the VM can live migrate to different node and keep previous 
+address keeping n/s and e/w communication.
+
+
+```text
+       ┌───────────────────────────┐
+       │     ovn_cluster_router    │
+┌───────────────────────────┐   ┌────────────────────────────┐
+│ static route              │───│ policy                     │
+│ prefix: 10.244.0.8        │   │ match: ip4.src==10.244.0.8 │
+│ nexthop: 10.244.0.8       │   │ action: reroute            │
+│ output-port: rtos-node1   │   │ nexthop: 10.64.0.2         │
+│ policy: dst-ip            │   │                            │  
+└───────────────────────────┘   └────────────────────────────┘ 
+```
+**Nodes logical switch ports:**
+
+The "router" logical switch port per node need to activate "arp_proxy" option
+with the VMs gateway and a fixed mac so VM neighbor cache is consistent 
+after live migration
+
+```text
+┌────────────────────┐     ┌────────────────────┐
+│logical switch node1│     │logical switch node2│
+┌────────────────────┐  │     │   ┌─────────────────────┐      
+│ lsp stor-node1     │──┘     └───│ lsp stor-node2      │      
+│ options:           │            │ options:            │    
+│  arp_proxy:        │            │   arp_proxy:        │
+│   0a:58:0a:f3:00:00│            │    0a:58:0a:f3:00:00│
+│   169.254.1.1      │            │    169.254.1.1      │
+└────────────────────┘            └─────────────────────┘
+```
+
+**VMs logical switch ports:**
+
+The logical switch port for new VMs will use ovn-k ipam to reserve an IP address,
+and when live-migrating the VM, the LSP address will be re-used.
+
+CNI must avoid setting the IP address on the migration destination pod.
+
```

Comment:
With label `kubevirt.io/allow-pod-network-bridge-binding-live-migration` ?

Reply:
We have an IPAM related epic for Openshift Virt that will also require this, but has **nothing** to do with the pod network.

If we do this based on a label / annotation, I would please ask you to use another label, whose name is more aligned to this. `kubevirt.io/skip-ip-configuration-on-cni` or something.

Reply:
..and if we are going this route, should this not be defined within kubevirt libs that we can import this label? 
Dont want to define it in OVN-K if at all possible. I assume its not defined there if we are discussing this here. @qinqon WDYT?

Reply:
> how will ovn-k know this pod is a kubevirt pod and not set IP in pods primary interface?

@martinkennelly , There is a chapter specific for that `Detecting hypershift VMs and live migration`, this is what ovn-k should do.


> We have an IPAM related epic for Openshift Virt that will also require this, but has **nothing** to do with the pod network.
> 
> If we do this based on a label / annotation, I would please ask you to use another label, whose name is more aligned to this. `kubevirt.io/skip-ip-configuration-on-cni` or something.

@maiqueb this annotation is not only used by ovn-k to activate DHCP options and disable CNI ip address configuration, it is also understood by [kubevirt](https://github.com/kubevirt/kubevirt/blob/e7016777cc6e8207d8dbc0cc9dcb47d83a79e83c/staging/src/kubevirt.io/api/core/v1/types.go#L938) to allow live migration with bridge binding, so calling this `skip-ip-configuration-on-cni` does not cover all what it does

Reply:
> ..and if we are going this route, should this not be defined within kubevirt libs that we can import this annotation? Dont want to define it in OVN-K if at all possible. I assume its not defined there if we are discussing this here. @qinqon WDYT?

@martinkennelly that is already defined there but I don't know if ovn-k maintainers will allow to import go dependency [`kubevirt.io/api`](https://github.com/kubevirt/api/blob/main/go.mod)

Reply:
why wouldnt we? I think its fine.

Reply:
Is the stuff under `kubevirt.io/api/go.mod` good for you ?  if this is the case is always better to import it.

Reply:
> That's what I'm advocating for: using two separate annotations, each with a different meaning.

@maiqueb I am fine with that, @martinkennelly @trozet are you ok have a pair of annotations instead of just one ? one for allow live migration with bridge binding at kubevirt and the other to skip setting ip configuration at CNI.

Reply:
@maiqueb maybe the annotation has to be `ovn.org/skip-ip-configuration-on-cni` instead of using the `kubevirt.io` prefix ?

Reply:
oh yes, I think that's better. Good catch.

Reply:
> Is the stuff under kubevirt.io/api/go.mod good for you ? if this is the case is always better to import it.

Implementation detail but looks fine :) No deps when importing it.

> I am fine with that, @martinkennelly @trozet are you ok have a pair of annotations instead of just one ? one for allow live migration with bridge binding at kubevirt and the other to skip setting ip configuration at CNI.

I am fine with it especially after Miguels context.

Reply:
> maybe the annotation has to be ovn.org/skip-ip-configuration-on-cni instead of using the kubevirt.io prefix ?

I was just about to say this! +++